File: LibraryCode/CPP/coinchange.cpp
/*
Coin Change

- given a list of coin sizings and a value V, determine the minimum
  number of coins needed to form said value V

- This algorithm assumes that it is guaranteed to be able to form V with
  the given coin sizings

Time: O(NV)
Space: O(NV)
*/
#include <bits/stdc++.h>  
using namespace std;
int N;               // Number of coin sizings  
int coins[101];      // Array of coin sizings
int memo[2000001];
// Remember to memset(memo, -1, sizeof memo);

// DP coinchange (True minimum)
// DP( x ) gives the minimum number of coins needed to form X
int DP (int price) {
    // Base Cases
    if (price == 0) return 0;
    if (price < 0) return 2e9;

    // Check if computed before
    int &ans = memo[price];
    if (ans != -1) return ans;

    // Find minimum number of coins by trying each coin, memoize state
    int minCoin = 2e9;
    for (int i = 0; i < N; i++) {
        minCoin = min(minCoin, 1 + DP(price - coins[i]));
    }
    return ans = minCoin;
}

File: prev/PY/coinchange.py
'''
O(NV) Coin Change (Complete search variant)

- DP returns the number of coins needed for each
  price

Time, Space: O(NV)
'''
import sys
from functools import lru_cache
sys.setrecursionlimit(int(1e9))

coins = [1,2,3,4,10]

# DP CoinChange (true minimum)
@lru_cache(maxsize=None)
def DP (price) :
    # Base cases
    if price == 0 : return 0
    if price < 0 : return 2e9
        
    minCoins = 2e9
    for coin in coins :
        minCoins = min(minCoins, 1 + DP(price - coin))
    return minCoins


print (DP(20))

